### PR TITLE
Add WhenTrait to Collection and MailMessage classes.

### PR DESCRIPTION
**Pull request by hegedustibor**
public function toMail($notifiable)
    {
         return (new MailMessage)
                    ->line('Welcome to the Laravel universe.')
                    ->attachWhen($this->it_can_attach_a_file, $this->attachment);
    }

This will attach a file to the message when the condition is true.

### CODE REVIEW

File: src/Illuminate/Notifications/Messages/MailMessage.php [hunk 1]
```diff
@@ -206,6 +206,19 @@ public function attach($file, array $options = [])
...
 
+    /**
+     * Attach a file to the message based on a given condition.
+     *
+     * @param bool $condition
+     * @param  string $file
+     * @param  array $options
+     * @return $this
+     */
+    public function attachWhen(bool $condition, $file, array $options = [])
```
Comment (xuanquynh) [inline src/Illuminate/Notifications/Messages/MailMessage.php]: Laravel doesn't allow primitive type-hint.

**Review by oliverquynh — CHANGES_REQUESTED** [review]

**Comment by crynobone** [comment]
would don't just have a universal `when($condition, callable $callback)` method.

File: src/Illuminate/Notifications/Messages/MailMessage.php (hunk 1 above)
Comment (hegedustibor) [reply src/Illuminate/Notifications/Messages/MailMessage.php]: Okay, thanks, I fixed it.

**Comment by taylorotwell** [comment]
Thanks for your pull request to Laravel!

Unfortunately, I'm going to delay merging this code for now. To preserve our ability to adequately maintain the framework, we need to be very careful regarding the amount of code we include. 

If possible, please consider releasing your code as a package so that the community can still take advantage of your contributions! 

If you feel absolutely certain that this code corrects a bug in the framework, please "**@**" mention me in a follow-up comment with further explanation so that GitHub will send me a notification of your response.

**Pull request closed without merging**